### PR TITLE
Implement two-factor authentication

### LINKED ISSUE #608: 2 Factor Auth support
This would be good to increase security when logging in.

### PR DESCRIPTION
**Pull request by tobias**
This provides optional two-factor authentication using TOTP codes. As part of that, this adds an internal event system to make it easier to send emails when events occur, and uses that system to email a user when 2fa is enabled/disabled.

I would like some review of this since it makes major changes to the authentication system.

The user flow is documented (somewhat) in the [wiki](https://github.com/clojars/clojars-web/wiki/Two-Factor-Auth).

This will resolve #608 and #727.

### CODE REVIEW

File: src/clojars/auth.clj [hunk 1]
```diff
@@ -23,3 +32,79 @@
     (f)
     (friend/throw-unauthorized friend/*identity*
       {:cemerick.friend/required-roles group})))
+
+(defn- get-param
+  [key form-params params]
+  (or (get form-params (name key)) (get params key "")))
```
Comment (pjstadig) [inline src/clojars/auth.clj:38]: Was it intentional to use `name` only once here?

Comment (tobias) [reply src/clojars/auth.clj:38]: This was adapted from https://github.com/cemerick/friend/blob/master/src/cemerick/friend/workflows.clj#L61, which looks up in `form-params` by string and in `params` by keyword. My assumption is that the ring middleware stack leaves form param keys as strings, but converts param keys to keywords. I suspect that the `(get form-params ,,,)` clause always returns the value we care about given the way we deliver the username/password/otp params from the login form, so the `or` clause isn't useful here.

**Comment by tobias** [comment]
Thanks @pjstadig!

File: src/clojars/auth.clj (hunk 1 above)
Comment (pjstadig) [reply src/clojars/auth.clj:38]: Cool. I figured something like that might be happening, but was not familiar enough to do anything but make sure it wasn't a typo! :)

File: src/clojars/web/login.clj [hunk 2]
```diff
@@ -34,5 +33,8 @@
              (password-field {:placeholder "keep it secret, keep it safe"
                               :required true}
                              :password)
+             (label :otp "Two-Factor Code")
+             (text-field {:placeholder "leave blank if two-factor auth not enabled"}
```
[the remark this answers is not in the document]
Comment (tobias) [reply src/clojars/web/login.clj:37]: I spent a little time trying to figure out how to do it in two stages, but but wasn't successful. But I would prefer this was two stages, so I may go back and try again to implement it before release.

File: test/clojars/integration/sessions_test.clj [hunk 3]
```diff
@@ -60,4 +62,38 @@
...
+            (has (status? 200))
+            (within [:.light-article :> :h1]
+                    (has (text? "Dashboard (fixture)")))))
+      (testing "with invalid token"
+        (-> (session app)
+            (login-as "fixture" "password" "1")
+            (follow-redirect)
+            (has (status? 200))
+            (within [:div :p.error]
+                    (has (text? "Incorrect username, password, or two-factor code.")))))
```
[the remark this answers is not in the document]
Comment (tobias) [reply test/clojars/integration/sessions_test.clj:85]: Will do.

**Pull request merged**